packages-exp/auth-exp/src/core/auth/firebase_internal.test.ts

```diff
@@ -0,0 +1,174 @@
+/**
+ * @license
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { expect } from 'chai';
+import * as sinon from 'sinon';
+
+import { testAuth, testUser } from '../../../test/helpers/mock_auth';
+import { Auth } from '../../model/auth';
+import { User } from '../../model/user';
+import { AuthInternal } from './firebase_internal';
+
+describe('src/core/auth/firebase_internal', () => {
+  let auth: Auth;
+  let authInternal: AuthInternal;
+  beforeEach(async () => {
+    auth = await testAuth();
+    authInternal = new AuthInternal(auth);
+  });
+
+  afterEach(() => {
+    sinon.restore();
+  });
+
+  context('getUid', () => {
+    it('returns null if currentUser is undefined', () => {
+      expect(authInternal.getUid()).to.be.null;
+    });
+
+    it('returns the uid of the user if set', async () => {
+      const user = testUser(auth, 'uid');
+      await auth.updateCurrentUser(user);
+      expect(authInternal.getUid()).to.eq('uid');
+    });
+  });
+
+  context('getToken', () => {
+    it('returns null if currentUser is undefined', async () => {
+      expect(await authInternal.getToken()).to.be.null;
+    });
+
+    it('returns the id token of the current user correctly', async () => {
+      const user = testUser(auth, 'uid');
+      await auth.updateCurrentUser(user);
+      user.stsTokenManager.accessToken = 'access-token';
+      user.stsTokenManager.expirationTime = Date.now() + 1000 * 60 * 60 * 24;
+      expect(await authInternal.getToken()).to.eql({
+        accessToken: 'access-token'
+      });
+    });
+  });
+
+  context('token listeners', () => {
+    let isProactiveRefresh = false;
+    let user: User;
+
+    beforeEach(async () => {
+      user = testUser(auth, 'uid', undefined, true);
+      await auth.updateCurrentUser(user);
+      sinon.stub(user.stsTokenManager, 'getToken').returns(
+        Promise.resolve({
+          accessToken: 'access-token',
+          refreshToken: 'refresh-tken',
+          wasRefreshed: true
+        })
+      );
+      sinon
+        .stub(user, '_startProactiveRefresh')
+        .callsFake(() => (isProactiveRefresh = true));
```

Comment:
can these just be sinon.spy()'s and you check to make sure they were called?

Reply:
I like this better since I'm testing as it gets toggled back and forth. I suppose you could do it by counting the number of times start was called and how many stop was called, but that seems more complicated